[Xcode] IDECustomBuildProductsPaths starting with "~" are ignored when determining the build location

emw-apple opened the pull request:
#### e8d345062f79e713b0e094449e83f8a0ce1bb5cc
<pre>
[Xcode] IDECustomBuildProductsPaths starting with &quot;~&quot; are ignored when determining the build location
<a href="https://bugs.webkit.org/show_bug.cgi?id=249442">https://bugs.webkit.org/show_bug.cgi?id=249442</a>
rdar://103460364

Reviewed by Alexey Proskuryakov.

On Apple ports, use bsd_glob to potentially expand a tilde in
user-specified path names. At some point in recent history, Xcode
appears to have stopped doing this automatically.

* Tools/Scripts/webkitdirs.pm:
(determineBaseProductDir):

Canonical link: <a href="https://commits.webkit.org/264450@main">https://commits.webkit.org/264450@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a25c75f9644115149de6f735b06355916d8bd6e7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7637 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7906 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8090 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9281 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7819 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7649 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9865 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7832 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10688 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7770 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/8877 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7037 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9391 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6185 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6958 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14643 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/6519 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7350 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7081 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10489 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/7228 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7574 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6182 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/7793 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6908 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1781 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1825 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11119 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/7999 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7305 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1924 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -412,7 +412,13 @@ sub determineBaseProductDir
+
+        # webkitdirs can be run from arbitrary directories, so any
+        # user-specified build directory must be absolute.
+        undef $baseProductDir unless defined($baseProductDir) && $baseProductDir =~ /^\//;
```
Is it appropriate to print warnings from `webkitdirs.pm`? I feel like I'd like to know if my settings are ignored here.

aproskuryakov commented on Tools/Scripts/webkitdirs.pm:
```diff
@@ -412,7 +412,13 @@ sub determineBaseProductDir
-        undef $baseProductDir unless $baseProductDir =~ /^\//;
+        # Expand tilde in pathnames for compatibility
+        # (https://bugs.webkit.org/show_bug.cgi?id=249442).
+        ($baseProductDir) = bsd_glob($baseProductDir, GLOB_TILDE) if $baseProductDir;
```
Probably best to check `if defined()` to not be at risk of hitting weird truthy values.

emw-apple committed:
[Xcode] IDECustomBuildProductsPaths starting with "~" are ignored when determining the build location
https://bugs.webkit.org/show_bug.cgi?id=249442
rdar://103460364

Reviewed by Alexey Proskuryakov.

On Apple ports, use bsd_glob to potentially expand a tilde in
user-specified path names. At some point in recent history, Xcode
appears to have stopped doing this automatically.

* Tools/Scripts/webkitdirs.pm:
(determineBaseProductDir):

Canonical link: https://commits.webkit.org/264450@main
